Update to Istio 1.0.0

rlenglet opened the pull request:
Upgrade the Envoy and Istio SHAs to the latest commit in https://github.com/istio/proxy/tree/1.0.0.
Update the generated Protobuf files.
Update the Cilium filters to Envoy's internal API changes.
Update the Vagrant box and cilium-builder Docker image to update the Bazel cache.  Install `ninja-build`, which is a new Envoy build dependency.
Update the proxy image dockerfiles to use Istio's new `proxytproxy` image.

Disable the ADS protocol support in the Cilium Agent, since it was unused.
Continuing to support ADS will require supporting Incremental xDS, cf. https://github.com/cilium/cilium/issues/5051.

Adapt Envoy route generation to Envoy's updated `HeaderMatcher` API.

Update the Istio GSG to Istio 1.0.0.
Enable the Pilot cilium plugin on the Pilot command line using the new `--plugins` flag.
Use the new Helm value to configure the Pilot image name.
Add `--validate=false` to every `kubectl` command to create deployments injected with Istio sidecar proxies, since kube-inject generates specs that are not 100% compatible with k8s 1.9.

Exit an endpoint's synchronous creation if it is in sidecar proxy mode.
If the endpoint is determined to have a sidecar proxy, return from the endpoint creation to let the sidecar container start, in case it is required to enforce L7 rules. Otherwise, endpoint creation would continuously timeout and the pod wouldn't start.

Signed-off-by: Romain Lenglet <romain@covalent.io>

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/4923)
<!-- Reviewable:end -->

rlenglet committed:
Update to Istio 1.0.0

Upgrade the Envoy and Istio SHAs to the latest commit in
https://github.com/istio/proxy/tree/1.0.0
Update the generated Protobuf files.
Update the Cilium filters to Envoy's internal API changes.

Update the Vagrant box and cilium-builder Docker image to update
the Bazel cache.  Install ninja-build, which is a new Envoy
build dependency.

Update the proxy image dockerfiles to use Istio's new proxytproxy
image.

Signed-off-by: Romain Lenglet <romain@covalent.io>

eloycoto commented on Vagrantfile:
```diff
@@ -127,7 +127,7 @@ Vagrant.configure(2) do |config|
 
         config.vm.box = "cilium/ubuntu-dev"
-        config.vm.box_version = "97"
+        config.vm.box_version = "100"
```
Quick one, if you can use 103 here will be great.

eloycoto commented on test/Vagrantfile:
```diff
@@ -8,7 +8,7 @@ $K8S_VERSION = ENV['K8S_VERSION'] || "1.10"
 $NFS = ENV['NFS']=="1"? true : false
 $SERVER_BOX = (ENV['SERVER_BOX'] || "cilium/ubuntu-dev")
-$SERVER_VERSION="97"
+$SERVER_VERSION="100"
```
Quick one, if you can use 103 here will be great.

rlenglet committed:
envoy: Disable unused ADS protocol support

Continuing to support ADS will require supporting Incremental
xDS, cf. https://github.com/cilium/cilium/issues/5051.

Signed-off-by: Romain Lenglet <romain@covalent.io>

rlenglet committed:
envoy: Adapt Envoy route generation to updated HeaderMatcher API

Signed-off-by: Romain Lenglet <romain@covalent.io>

rlenglet committed:
docs: Update the Istio GSG to Istio 1.0.0

Enable the Pilot cilium plugin on the Pilot command line using the
new --plugins flag.
Use the new Helm value to configure the Pilot image name.

Add --validate=false to every kubectl command to create deployments
injected with Istio sidecar proxies, since kube-inject generates
specs that are not 100% compatible with k8s 1.9.

Signed-off-by: Romain Lenglet <romain@covalent.io>

rlenglet committed:
daemon: Exit synced endpoint creation if in sidecar proxy mode

If the endpoint is determined to have a sidecar proxy, return
from the endpoint creation to let the sidecar container start,
in case it is required to enforce L7 rules.
Otherwise, endpoint creation would continuously timeout and the
pod wouldn't start.

Signed-off-by: Romain Lenglet <romain@covalent.io>

raybejjani commented on daemon/endpoint.go:
```diff
@@ -259,8 +259,15 @@ func (h *putEndpointID) Handle(params PutEndpointIDParams) middleware.Responder
+					// If the endpoint is determined to have a sidecar proxy,
+					// return immediately to let the sidecar container start,
+					// in case it is required to enforce L7 rules.
+					return NewPutEndpointIDCreated()
```
Could you add a log here explaining the early return? I would advocate for an Info, since it's violating the synchronous wait flag.

Is there no other place to make this check? the `e.hasSidecarProxy` flag seems to be based only on the labels from k8s. Is it already available when `cilium-cni` is invoked? If so, could we then not use the synchronous option in that case?

rlenglet replied:
Added an info log statement.

Yes, the sidecar proxy mode is determined based on k8s labels. Since the synchronous request will wait for up to 55 seconds, it gives enough time for the daemon to update the endpoint's labels. It typically happens within a few seconds after the endpoint creation, in a separate controller. So before the synchronous request times out, we'll know whether the endpoint is in sidecar mode.

raybejjani replied:
Oh, I see. I thought these labels were set on endpoint create. If we wait on the labels from k8s I see how this would wait until they arrive, which will be some time after the endpoint create. I guess in that sense it isn't really violating the synchronous flag since it still waits.

raybejjani commented on pkg/envoy/sort.go:
```diff
@@ -214,17 +214,75 @@ func HeaderMatcherLess(m1, m2 *envoy_api_v2_route.HeaderMatcher) bool {
 		return false
```
Since this function is now quite a bit more complex, could you add a comment explaining the general logic? The getters are autogenerated and I'm not sure how to figure out what they fields are actually for.

rlenglet replied:
Added this comment:
```
	// Compare the header_match_specifier oneof field, by comparing each
	// possible field in the oneof individually:
	// - exactMatch
	// - regexMatch
	// - rangeMatch
	// - presentMatch
	// - prefixMatch
	// - suffixMatch
	// Use the getters to access the fields and return zero values when they
	// are not set.
```

raybejjani replied:
Excellent, exactly what I needed :) Thanks

raybejjani reviewed (dismissed):
Mostly fine. Neither of my comments is particularly blocking, in case you want to merge this before I get a chance to approve.
